File: python/euler/bin/p0046.py
# project euler: problem 46

from itertools import count
from math import isqrt

from euler.lib.prime import is_prime


def is_twice_squre(n: int) -> bool:
    tmp = isqrt(n // 2)

    return n % 2 == 0 and tmp * tmp == n // 2


def compute() -> str:
    # Two is not an odd number.
    odd_primes = [3, 5, 7, 11, 13, 17, 19, 23, 29, 31]

    # The odd composite numbers less than 35 have been written in the problem statement.
    for x in count(35, 2):
        if is_prime(x):
            odd_primes.append(x)
        elif any(is_twice_squre(x - p) for p in odd_primes):
            pass
        else:
            return str(x)

    assert False, 'unreachable!'


def solve() -> str:
    return compute()

File: typescript/euler/bin/p0046.ts
// project euler: problem 46

import { isqrt } from "../lib/math.ts";
import { isPrimeSimple } from "../lib/primes.ts";

const isTwiceSquare = (n: number): boolean => {
  const tmp = Math.trunc(n / 2);
  const tmp_isqrt = isqrt(tmp);

  return n % 2 === 0 && tmp_isqrt * tmp_isqrt === tmp;
};

export const compute = (): string => {
  // '2' is not an odd prime number
  const oddPrimes = [3, 5, 7, 11, 13, 17, 19, 23, 29, 31];

  // The odd composite numbers less than 35 have been written in the problm statement
  loop_x:
  for (let x = 35;; x += 2) {
    if (isPrimeSimple(x) === true) {
      oddPrimes.push(x);
      continue;
    }
    for (const p of oddPrimes) {
      if (isTwiceSquare(x - p) === true) {
        continue loop_x;
      }
    }
    return String(x);
  }
};

export const solve = (): string => compute();
